### PR TITLE
Implements CntProbe

### PR DESCRIPTION
**Pull request by SmallRoomLabs**
I realized that I've got a lot of counters and probes for debugging setup all over the place in my PDP8 design. So I made this "counter probe" that can replace all of those.  Smaller and easer&faster to place where needed.

It took me a good 5-6 hours to figure out how to get this working - but it gave me a deeper insight on how Digital is designed so I consider it well spent time.  I'm starting to think that it's kinda fun to make components - hopefully it won't become an obsession. ^__^

Screen Shot 2019-10-14 at 2 11 57 PM

Just to be perfectly clear - the "#" thingie at the bottom is the CntProbe, the top part is how I used to do it.

Please have a quick peek at it and see if I've messed up something somewhere.

### CODE REVIEW

**Comment by SmallRoomLabs** [comment]
Hmmm... I see that I've called the branch CntPort instead of CntProbe - the two terms are so similar in spelling that my muscle memory triggered and fsckd it up.  Not that it *really* matters though...

**Comment by hneemann** [comment]
Looks ok. But there is one detail which may behave incorect if the clock frequenc is high.
Every shape has a method `public void readObservableValues()` this method needs to be overwritten and used like this:

```
    @Override
    public void readObservableValues() {
        if (cntProbe != null)
            cntValue = cntProbe.getCnt();
    }

    @Override
    public void drawTo(Graphic graphic, Style highLight) {
        int dy = -1;
        Orientation orientation = Orientation.LEFTCENTER;
        if (isLabel) {
            graphic.drawText(new Vector(2, -4), new Vector(3, -4), label, Orientation.LEFTBOTTOM, Style.NORMAL);
            dy = 4;
            orientation = Orientation.LEFTTOP;
        }
        String v = "#";
        if (cntProbe != null)
            v = Integer.toString(cntValue);
        graphic.drawText(new Vector(2, dy), new Vector(3, dy), v, orientation, Style.NORMAL);
    }
```

And that is the reason: Drawing the circuit takes a very long time. You can't stop the simulation that long without reducing the speed a lot. But if you don't stop the simulation, an inconsistent drawing can occur because the simulation evolves during the drawing.
The solution is to divide the drawing process into a part that is processed very quickly and only remembers the state to be drawn. During this time the simulation is stopped. Then the simulation is continued and the drawing is created while the simulation is running.
In short: In the `drawTo()` method the model may not be queried. This is only allowed in the `readObservableValues()` method. 
And the `readObservableValues()` method should do as little as possible, to return as soon as possible.

**Comment by SmallRoomLabs** [comment]
Ah, ok, that was an easy fix.  Thanks.

But I actually thought earlier about the separate threads for calculating and drawing and that they would not be in total sync with eachother but dismissed that as a problem since they would eventually reach a consistent state.   But I guess you only repaint dirty wires and objects and then some of the erroneous states might be there basically forever.

By the way; how big a difference in speed is there between a subcircuit and a native component implementing the same function?  If we for instance take something very simple like this edge detector?  Would it affect the simulation in any measurable way?

Screen Shot 2019-10-14 at 5 55 18 PM

**Comment by hneemann** [comment]
> But I actually thought earlier about the separate threads for calculating and drawing and that they would not be in total sync with eachother but dismissed that as a problem since they would eventually reach a consistent state. But I guess you only repaint dirty wires and objects and then some of the erroneous states might be there basically forever.

This type of race condition has no influence on the simulation itself. It only leads to an inconsistent graphical representation because parts of the circuit are drawn in a different simulation state than others.

> By the way; how big a difference in speed is there between a subcircuit and a native component implementing the same function? If we for instance take something very simple like this edge detector? Would it affect the simulation in any measurable amount so it would be worth the effort?

I don't think that makes a measurable difference. With a more complex circuit, it would probably be noticeable, but not with this simple circuit.

**Comment by SmallRoomLabs** [comment]
With "reach a consistent state" I referred to the graphical state in regards to the logic state not the logic state itself.   For a sim with a clock >10Hz then it probably doesn't matter I guess, but for for a slow clock or single step it's another matter.

**Comment by SmallRoomLabs** [comment]
Then I won't implement the edge detectors as components.  They are small and unobtrusive enough as a subcircuit.

Screen Shot 2019-10-14 at 6 57 16 PM

So back to the final testing and debugging of the PDP8.  And also to write some scripts that can automatically apply my own components  to the source whenever you release a new version.

**Comment by hneemann** [comment]
> And also to write some scripts that can automatically apply my own components to the source whenever you release a new version.

Have you ever thought about creating a plug-in for Digital? Then you can develop your own components more easily and don't depend on the complete source code.
See this [example](https://github.com/hneemann/digitalCustomComponents).

**Comment by SmallRoomLabs** [comment]
Ohh, that's nice!  Would both the CntProbe- and the Port- components be possible to do with the plugin-architecture?

**Comment by hneemann** [comment]
I think so. If not, I'll add what's missing.

**Comment by SmallRoomLabs** [comment]
Ollright, I'll give it a whirl.

**Comment by SmallRoomLabs** [comment]
There's no need to keep this open, so I'll close it now.
